[Exam] Variants: Data model

sascha11110 opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context

In order to be able to use Artemis for conducting exams, some adjustments to the data model are required.

### Description

- [x] We've implemented all new entities
- [x] We've added DTO classes for all new entities
- [x] We've created the corresponding Liquibase change-log
- [x] We've added corresponding TypeScript classes in the client
- [x] We've clarified everything that has *to be discussed*
- [x] We've added `extends AbstractAuditingEntity` to all new entities so that `created_date`, `last_modified`, ... are added to the database rows

![Exam Mode](https://user-images.githubusercontent.com/15168372/84052944-87bf7b00-a9b1-11ea-949c-e894081c8bd1.png)

<!--
### To be discussed

#### 1) Bidirectional mappings
Some of the bidirectional mappings are not yet implemented. Should we add the following mappings?

- **`StudentExam` <-> `User`:** should `Set<StudentExam>` be added to `User`?
- **`ExerciseGroup` <-> `Exercise`:** should `ExerciseGroup` be added to `Exercise`?
- **`StudentExam` <-> `Exercise`:** should `Set<StudentExam>` be added to `Exercise`?

#### 2) DTOs
Some of the attributes of the `Exam` entity should sometimes not be sent to the client (e.g. `maxScore`). @krusche suggested to add another class `ExamConfig` that has a one to one relation to `Exam` to make sure that something like `maxScore` is not sent accidentally. We'd like to use DTOs for these cases.

At the moment we define a default serialization with Jackson and set attributes to null in a resource or service if we don't want to send them over the network. DTOs provide a more consistent but also flexible way to define what should and what shouldn't be sent.

We've added an exemplary DTO implementation for the `Exam` entity (see `ExamDTO` and `ExamInstructorDTO`). Should we add DTOs for all new entities?

#### 3) `@Lob`/`LONGTEXT` vs `MediumText`
For descriptions the `@Lob` annotation is used (`LONGTEXT` in MySql) that might be a potential security vulnerability due to its size. Therefore we decided to use `String` in the Java code and [`MEDIUMTEXT`](https://dev.mysql.com/doc/refman/8.0/en/blob.html) in the database.

#### 4) `numberOfExercisesPerStudent` in `Exam`
What do we need this attribute for?
-->

## Additional changes
If exams are available, they are now part of the REST Call `/courses/for-dashboard`. We optimized the performance here, so please double check that we did not introduce unwanted behavior. 


## Steps for Testing
1. Login
2. Open the course overview, course details and exercises
3. Check that everything is displayed correctly (please compare this to the current state in `develop`

krusche commented:
Thanks for the fast creation of this PR and your initiative here. I really appreciate it!

Here are my thoughts on your discussion points:

1) Bidirectional mappings
* StudentExam <-> User: No
* ExerciseGroup <-> Exercise: No
* StudentExam <-> Exercise: No

I would try out to work without bidirectional mappings for now. We could add them later on, in case they are needed

2) We can try out DTOs if they make things easier here.

3) Why would this be a security issue? I guess MediumText is fine.

4) We can rename this if it's not clear. Let's assume there are 6 exercise groups in one exam, but the instructor only wants to have 4 exercises in the exam. Then `numberOfExercisesPerStudent` (or e.g. renamed to `numberOfExercisesInExam`) would be 4. During the assignment (see below), this would be taken into account


I think we need at least 2 actions in the exam

1) **Register students**: upload a CSV file, it should also be possible to add/remove single students, we can reuse the functionality of the course groups here, maybe @madwau can help here, because he has programmed this --> **this also means that the Exam class needs a many to many association to User** (unidirectional again) so that we can store which students are registered for an exam. As we get this information from external systems (in this case TUM Online), students cannot register on their own.

2) **Assign exercise variants**: This will compute the student exams based on the preconfigured options for the exam and the defined variants in the exercise groups

Some considerations for exam:
- instructors should be able to specify a visibleDate (this needs to be before the start date and defines when the exam mode will be visible to students, e.g. 5 min before the start date)
- releaseDate should be renamed to startDate
- dueDate should be renamed to endDate


Please further discuss which attributes we might need.

sascha11110 commented:
@krusche Thanks for the quick feedback! We've added some attributes and I've updated the diagram in the PR description in accordance to your comment.

1. We've added a `title` to `Exam`.
2. We've added the `visibleDate`.
3. We've changed `releaseDate` and `dueDate` to `startDate` and `endDate`.
4. We've added `numberOfExercisesPerStudent` as **`numberOfExerciseGroupsPerStudent`**.
5. We've added the unidirectional mapping between `Exam` and `User`.
6. **To be discussed:** We've added `creationDate` to `StudentExam`. It might be relevant when the student exam was generated.
7. **To be discussed:** We've added `position` to `ExerciseGroup` as we think that the order of the exercise (groups) is relevant for the exams.
8. **To be discussed:** Do we really need the `maxScore` in `Exam`? Wouldn't it be less error-prone to derive the maximum score from the associated exercise groups? (btw: we or at least the instructor should ensure that all combinations of optional and mandatory exercise groups lead to the same maximum score)

> Why would this be a security issue? I guess MediumText is fine.

Admittedly, this was more a theoretical consideration, which is more relevant for text fields that can also be filled by non-instructors (or non-admins). `@Lob` with `LARGETEXT` means that someone can write up to 4 GB of text into the database. However, I think that we already limit the number of characters in relevant places (e.g. Q&A).

stefanwaldhauser commented:
I reworked the entities  and included all relevant methods for adding/removing parts of a relatonship that make sure the relationships are consistent. The next logical step would be to set up some Request/Response DTOs, the repositories and some basic controllers imo.

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+     * From all exercise groups connected to the exam, {@link #numberOfExerciseGroups} are randomly
+     * chosen when generating the specific exam for the {@link #registeredUsers}
+     */
+    @Column(name = "NUMBER_OF_EXERCISE_GROUPS")
```
```suggestion
    @Column(name = "number_of_exercise_groups")
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+    // -----------------------------------------------------------------------------------------------------------------
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "COURSE_ID")
```
```suggestion
    @JoinColumn(name = "course_id")
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+
+    // -----------------------------------------------------------------------------------------------------------------
+    @ManyToMany
+    @JoinTable(name = "EXAM_JHI_USER", joinColumns = @JoinColumn(name = "EXAM_ID", referencedColumnName = "ID"), inverseJoinColumns = @JoinColumn(name = "STUDENT_ID", referencedColumnName = "ID"))
```
```suggestion
    @JoinTable(name = "exam_jhi_user", joinColumns = @JoinColumn(name = "exam_id", referencedColumnName = "id"), inverseJoinColumns = @JoinColumn(name = "student_id", referencedColumnName = "id"))
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+import de.tum.in.www1.artemis.domain.User;
+
+@Entity
+@Table(name = "STUDENT_EXAM")
```
```suggestion
@Table(name = "student_exam")
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+    // -----------------------------------------------------------------------------------------------------------------
+
+    @Id
+    @Column(name = "ID")
```
```suggestion
    @Column(name = "id")
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+    // -----------------------------------------------------------------------------------------------------------------
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "EXAM_ID")
```
```suggestion
    @JoinColumn(name = "exam_id")
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+
+    // -----------------------------------------------------------------------------------------------------------------
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "USER_ID")
```
```suggestion
    @JoinColumn(name = "user_id")
```

JonasPetry commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+
+    // -----------------------------------------------------------------------------------------------------------------
+    @ManyToMany
+    @JoinTable(name = "STUDENT_EXAM_EXERCISE", joinColumns = @JoinColumn(name = "STUDENT_EXAM_ID", referencedColumnName = "ID"), inverseJoinColumns = @JoinColumn(name = "EXERCISE_ID", referencedColumnName = "ID"))
```
```suggestion
    @JoinTable(name = "student_exam_exercise", joinColumns = @JoinColumn(name = "student_exam_id", referencedColumnName = "id"), inverseJoinColumns = @JoinColumn(name = "exercise_id", referencedColumnName = "id"))
```

stefanwaldhauser commented:
> I am not sure why all column names are referenced in caps. I suggest to change them to lower case.

"The SQL standard requires identifiers be stored in all-uppercase. See section 5.2.13 of the SQL-92 as quoted from a draft copy in this Answer on another Question. The standard allows you use undelimited identifiers in lowercase or mixed case, as the SQL processor is required to convert as needed to convert to the uppercase version."

I know it will be converted to uppercase internally anyway, but by convention you should always use Uppercase

JonasPetry commented:
> > I am not sure why all column names are referenced in caps. I suggest to change them to lower case.
> 
> "The SQL standard requires identifiers be stored in all-uppercase. See section 5.2.13 of the SQL-92 as quoted from a draft copy in this Answer on another Question. The standard allows you use undelimited identifiers in lowercase or mixed case, as the SQL processor is required to convert as needed to convert to the uppercase version."
> 
> https://stackoverflow.com/questions/2029358/should-i-write-table-and-column-names-always-lower-case#:~:text=SQL%20standard%20requires%20names%20stored%20in%20uppercase&text=13%20of%20the%20SQL%2D92,convert%20to%20the%20uppercase%20version.
> 
> if our convention is to use lower case for everything, i will change it.

I'm not sure what our convention is tbh, but it is written in lowercase in most of the other cases. Maybe we want to change that though. What is your call on this @krusche ?

stefanwaldhauser commented:
> > > I am not sure why all column names are referenced in caps. I suggest to change them to lower case.
> > 
> > 
> > "The SQL standard requires identifiers be stored in all-uppercase. See section 5.2.13 of the SQL-92 as quoted from a draft copy in this Answer on another Question. The standard allows you use undelimited identifiers in lowercase or mixed case, as the SQL processor is required to convert as needed to convert to the uppercase version."
> > https://stackoverflow.com/questions/2029358/should-i-write-table-and-column-names-always-lower-case#:~:text=SQL%20standard%20requires%20names%20stored%20in%20uppercase&text=13%20of%20the%20SQL%2D92,convert%20to%20the%20uppercase%20version.
> > if our convention is to use lower case for everything, i will change it.
> 
> I'm not sure what our convention is tbh, but it is written in lowercase in most of the other cases. Maybe we want to change that though. What is your call on this @krusche ?

I think it is done lower case everywhere. I will change it :)

krusche commented:
Thanks for the changes. I will add some feedback directly in the code

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:
```diff
@@ -153,6 +191,15 @@
+
+    @ManyToMany(mappedBy = "exercises")
+    @JsonIgnore
+    private Set<StudentExam> studentExams = new HashSet<>();
```
please leave this out for now, because we don't need this reference to be bidirectional

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/ExerciseGroup.java:
```diff
@@ -0,0 +1,174 @@
+
+@Entity
+@Table(name = "exercise_group")
+public class ExerciseGroup extends AbstractAuditingEntity {
```
why is it an abstract auditing entity? should not be necessary!

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/ExerciseGroup.java:
```diff
@@ -0,0 +1,174 @@
+    public ExerciseGroup() {
+    }
+
+    public ExerciseGroup(Long id, String title, Boolean isMandatory, Integer positionInExam, Exam exam, Set<Exercise> exercises) {
```
please do not use constructors like this

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/ExerciseGroup.java:
```diff
@@ -0,0 +1,174 @@
+     * Specifies the position of this exercise group if it is used in an exam
+     */
+    @Column(name = "position_in_exam", nullable = false)
+    private Integer positionInExam = 1;
```
please remove this attribute. We should rather use an ordered collection for Exam --> ExerciseGroup

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/User.java:
```diff
@@ -117,6 +136,14 @@
 
+    @ManyToMany(mappedBy = "registeredUsers")
+    @JsonIgnore
+    private Set<Exam> exams = new HashSet<>();
```
Please remove this. It won't be necessary

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/User.java:
```diff
@@ -117,6 +136,14 @@
+
+    @OneToMany(mappedBy = "user")
+    @JsonIgnore
+    private Set<StudentExam> studentExams = new HashSet<>();
```
Please remove this. It won't be necessary

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+    public Exam() {
+    }
+
+    public Exam(Long id, String title, ZonedDateTime visibleDate, ZonedDateTime startDate, ZonedDateTime endDate, String startText, String endText, String confirmationStartText,
```
I would try to avoid the use of constructors

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+
+@Entity
+@Table(name = "exam")
+public class Exam extends AbstractAuditingEntity {
```
why is it an abstract auditing entity? should not be necessary

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+
+    // -----------------------------------------------------------------------------------------------------------------
+    @OneToMany(mappedBy = "exam")
+    private Set<ExerciseGroup> exerciseGroups = new HashSet<>();
```
please use a List and OrderedColumn here

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+        if (o == null || getClass() != o.getClass())
+            return false;
+        Exam exam = (Exam) o;
+        return Objects.equals(title, exam.title);
```
I think this is wrong. We should use the id here!

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(title);
```
please use id

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id);
```
```suggestion
        return Objects.hashCode(getId());
```

krusche replied:
Please change this accordingly in all new entities

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+        if (o == null || getClass() != o.getClass())
+            return false;
+        StudentExam that = (StudentExam) o;
+        return Objects.equals(id, that.id);
```
```suggestion
        return Objects.equals(getId(), that.getId());
```

krusche replied:
Please change this accordingly in all new entities

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+
+@Entity
+@Table(name = "student_exam")
+public class StudentExam extends AbstractAuditingEntity {
```
why do we have AbstractAuditingEntity here? I don't think this is necessary

sascha11110 replied:
We thought that it might be interesting to know when the student's exam was created and when it was updated last time. You think we shouldn't track this here? 

I've removed `AbstractAuditingEntity` from `Exam` and `ExerciseGroup`.

krusche replied:
I imagine that they are only generated once. In case a 2nd Generation would be necessary, we would delete all existing objects and then recreate them based on the new config. Then we would have no use for auditing

sascha11110 replied:
Okay, I will remove it.

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+    public StudentExam() {
+    }
+
+    public StudentExam(Long id, Exam exam, User user, Set<Exercise> exercises) {
```
I would avoid the use of constructors

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+
+    public void setExam(Exam exam) {
+        if (this.exam != null) {
+            this.exam.removeStudentExam(this);
```
I would be careful with additional logic like this. Please have a look how this is handled in other places in our code, so that we do not break the consistency

sascha11110 replied:
We simply do not implement the additional logic in the entities. E.g. in `Submission`:

```
public void setParticipation(Participation participation) {
        this.participation = participation;
}
```

Therefore I'll remove it in the new entities.

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:
```diff
@@ -0,0 +1,154 @@
+
+    public void setUser(User user) {
+        if (this.exam != null) {
+            this.exam.removeStudentExam(this);
```
same comment as above

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/dto/converter/ExamDTOConverter.java:
```diff
@@ -0,0 +1,17 @@
+import de.tum.in.www1.artemis.domain.exam.dto.ExamDTO;
+
+@Component
+public class ExamDTOConverter implements Converter<Exam, ExamDTO> {
```
we do not use such converters until now in the code. What would be the advantage?

sascha11110 replied:
I think we do not really need the converters so I removed them. They were introduced to enable thread-safe conversion.

Nevertheless, we should define a way to convert entities to DTOs and vice versa. I think its not good practice to add the conversion methods to the entities and DTOs itself. We might come up with our own converter classes or with a library like [modelmapper](http://modelmapper.org/).

It would be nice to have something like the following: 
```
@PostMapping
public void createExam(@DTO(ExamInstructorDTO.class) Exam exam) {
     examRepository.save(exam);
}
```

krusche commented on src/main/webapp/app/entities/exam.model.ts:
```diff
@@ -0,0 +1,18 @@
+import { StudentExam } from 'app/entities/student-exam.model';
+import { ExerciseGroup } from 'app/entities/exercise-group.model';
+
+export class Exam {
```
```suggestion
export class Exam implements BaseEntity {
```

krusche replied:
all client entities should implement BaseEntity

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+     * chosen when generating the specific exam for the {@link #registeredUsers}
+     */
+    @Column(name = "number_of_exercise_groups")
+    private Integer numberOfExerciseGroups;
```
Can we rename this to `numberOfExercisesInExam`

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:
```diff
@@ -0,0 +1,312 @@
+
+    @Column(name = "confirmation_end_text")
+    private String confirmationEndText;
+
```
please add `maxPoints` as Integer.
Instructors should define this (e.g. 60) and this should be independent of the exercise groups or exercises.
We can use this as additional validation constraint later on, so that we could display a warning if the instructor tries to generate student exams, but the configuration does not fit.

Keep in mind: each student exam should have the same number of points. This is also important when it comes to the calculation of the grades, which we might support at a later stage directly in Artemis

sascha11110 commented:
@krusche I've addressed all of your change requests and updated the diagram in the description accordingly. I've created a first version of the liquibase schema but I will have a closer look at it later.

**To be discussed:**

1. Open conversation: https://github.com/ls1intum/Artemis/pull/1581#discussion_r436391404
2. We have a relationship between `Exercise` in `StudentExam`. Should `Set<Exercise> exercises` in `StudentExam` also be an ordered `List`?

krusche commented:
I think we need two more small changes:
1) Instructors should be able to specify in the exam whether the order of exercises for the randomly generated student exams should have an order (boolean flag)
2) To support an order, we need to store the reference `StudentExam -> Exercise` in an ordered list

stefanwaldhauser commented:
> I think we need two more small changes:
> 
> 1. Instructors should be able to specify in the exam whether the order of exercises for the randomly generated student exams should have an order (boolean flag)
> 2. To support an order, we need to store the reference `StudentExam -> Exercise` in an ordered list

Just to get it right, instructors should be able to select that the exercise groups (the exercises inside of each exercise group are chosen randomly) are presented to the student in a fixed order? This would mean that we need to store the specific order somewhere. Originally we have decided to specify it as a field in exerciseGroup but I see that it was removed now.

krusche commented:
The order is already stored, because it's an ordered list as relationship

stefanwaldhauser commented:
> The order is already stored, because it's an ordered list as relationship

I understand this on the Java Code level, but when the instructor creates the exam layout, this information will somehow need to be stored in the database, no? There is something like  @Ordered Column in JPA https://en.wikibooks.org/wiki/Java_Persistence/Relationships#Order_Column_.28JPA_2.0.29in
but you need a column in the database for this

krusche commented:
> > The order is already stored, because it's an ordered list as relationship
> 
> I have added the annotation "@OrderColumn(name = "exercise_group_order")". This will add the necessary database columns to keep the ordering.
> https://en.wikibooks.org/wiki/Java_Persistence/Relationships#Order_Column_(JPA_2.0)

Yes, exactly. The ordered column makes sure the order is persisted. It basically adds a new column for the position in the list. Have a look at how this is done for Result --> Feedback.

krusche commented:
I added some first REST calls.
Also the courses for dashboard call now includes exams in a performant way so that they can be displayed for users in case they exist. This will be part of the exam mode team (who focuses on the student functionality on the client side) though.
